File: heap/heap.test.c
#include "heap.h"

#include <string.h>

#include "tester.h"

#define matches(a, b) (!strcmp(a,b))

const char *TesterHelpStrings[] = {
    "This is the heap test module",
    "This implements a min heap, meaning that the first item extracted",
    "is the smallest of all in the heap",
    "",
    "/create <size>               create a heap with size <size>",
    "/insert <number>             insert <number> in the heap",
    "/extract <expected number>   extract number from heap",
    "/resize <new size>           resize heap to size <new size>",
    NULL,
};

static Heap *pHeap;

TesterReturnValue TesterInitCallback()
{
    pHeap = NULL;
    return TESTER_RETURN_OK;
}

TesterReturnValue convert(HeapReturnID heapReturnValue)
{
    switch (heapReturnValue) {
    case HEAP_RETURN_OK:
        return TESTER_RETURN_OK;
    case HEAP_RETURN_EMPTY:
        return TesterExternalReturnValue("empty");
    case HEAP_RETURN_FULL:
        return TesterExternalReturnValue("full");
    case HEAP_RETURN_SHRINK:
        return TesterExternalReturnValue("shrink");
    case HEAP_RETURN_INVALID_PARAMETER:
        return TesterExternalReturnValue("invalid parameter");
    case HEAP_RETURN_MEMORY:
        return TesterExternalReturnValue("memory");
    default:
        return TesterExternalReturnValue("unknown");
    }
}

int cmpObjs(void *a, void *b, void *_unused)
{
    return *((int *) a) < *((int *) b); /* MIN HEAP */
}

TesterReturnValue TesterParseCallback(const char *command)
{
    HeapReturnID returnId = HEAP_RETURN_OK;
    if matches(command, "create") {
        size_t size;
        Heap *temp;
        if (TesterParseArguments("z", &size) != 1)
            return TESTER_RETURN_ARGUMENT;
        if (pHeap != NULL)
            heapDestroy(pHeap);
        returnId = heapCreate(&temp, size, cmpObjs, free, NULL);
        if (!returnId)
            pHeap = temp;
    } else if matches(command, "insert") {
        int obj, *pObj;
        if (TesterParseArguments("i", &obj) != 1)
            return TESTER_RETURN_ARGUMENT;
        pObj = malloc(sizeof(int));
        if (!pObj)
            return TESTER_RETURN_MALLOC;
        *pObj = obj;
        returnId = heapInsert(pHeap, pObj);
        if (returnId)
            free(pObj);
    } else if matches(command, "extract") {
        int *pObj, actual, expected;
        if (TesterParseArguments("i", &expected) != 1)
            return TESTER_RETURN_ARGUMENT;
        returnId = heapExtract(pHeap, &pObj);
        if (!returnId) {
            actual = *pObj;
            free(pObj);
            if (actual != expected)
                return TESTER_RETURN_RETURN;
        }
    } else if matches(command, "resize") {
        size_t newSize;
        if (TesterParseArguments("z", &newSize) != 1)
            return TESTER_RETURN_ARGUMENT;
        returnId = heapResize(pHeap, newSize);
    } else {
        return TESTER_RETURN_COMMAND;
    }
    return convert(returnId);
}

TesterReturnValue TesterExitCallback()
{
    if (pHeap) {
        heapDestroy(pHeap);
        pHeap = NULL;
    }
    return TESTER_RETURN_OK;
}

File: heap/python/heap_test.py
from pyheap import *

def test_create():
	h = Heap()
	h = Heap(size=127)
	assert h.size() == 127
	def func(a,b):
		return a > b
	h = Heap(f=func)
	h = Heap(f=func, size=127)
	assert h.size() == 127

def test_full():
	h = Heap(size=1)
	h.insert(1)
	try:
		h.insert(2)
	except Full:
		return
	assert False

def test_empty():
	h = Heap()
	try:
		h.extract()
	except Empty:
		return
	assert False

def test_lock():
	from random import random
	l = []
	def evil(a,b):
		l[0].extract()
		return a < b
	h = Heap(f=evil, size=7)
	l.append(h)
	try:
		for i in range(7):
			h.insert(random())
		for i in range(7):
			print(h.extract())
	except Lock:
		return
	assert False

def test_shrink():
	h = Heap(size=15)
	for i in range(7):
		h.insert(i)
	h.resize(7)
	assert h.size() == 7
	try:
		h.resize(3)
	except Shrink:
		assert h.size() == 7
		return
	assert False

def test_ordering():
	from random import random
	def minheap(a,b):
		return a < b
	def maxheap(a,b):
		return a > b
	for func in [minheap, maxheap]:
		h = Heap(f=func, size=100)
		for i in range(100):
			h.insert(random())
		prev = h.extract()
		for i in range(99):
			new = h.extract()
			assert func(prev,new)
			prev = new